state/unitstate.go

```diff
@@ -0,0 +1,243 @@
+// BSON does not allow ints as a map key.
+func (d *unitStateDoc) relationData() (map[int]string, error) {
+	if d.RelationState == nil {
+		return nil, nil
+	}
+	// BSON maps cannot have an int as key.
+	rState := make(map[int]string, len(d.RelationState))
+	for k, v := range d.RelationState {
+		kString, err := strconv.Atoi(k)
+		if err != nil {
+			return nil, err
+		}
+		rState[kString] = v
+	}
+	return rState, nil
+}
+
+// relationStateMatches returns true if the RelationState map within the
+// unitStateDoc matches the provided newRS argument.  Assumes that
+// IgnoreRelationsState has been called first, therefore if arg is empty,
+// returns a nil map to be set.
+func (d *unitStateDoc) relationStateMatches(newRS map[string]string) bool {
+	for k, v := range d.RelationState {
+		if newRS[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// removeUnitStateOp returns the operation needed to remove the unit state
+// document associated with the given globalKey.
+func removeUnitStateOp(mb modelBackend, globalKey string) txn.Op {
+	return txn.Op{
+		C:      unitStatesC,
+		Id:     mb.docID(globalKey),
+		Remove: true,
+	}
+}
+
+// UnitState contains the various state saved for this unit,
+// including from the charm itself and the uniter.
+type UnitState struct {
+	// state encodes the unit's persisted state as a list of key-value pairs.
+	state    map[string]string
+	stateSet bool
+
+	// uniterState is a serialized yaml string containing the uniters internal
+	// state for this unit.
+	uniterState    string
+	uniterStateSet bool
+
+	// relationState is a serialized yaml string containing relation internal
+	// state for this unit from the uniter.
+	relationState    map[int]string
+	relationStateSet bool
+
+	// storageState is a serialized yaml string containing storage internal
+	// state for this unit from the uniter.
+	storageState    string
+	storageStateSet bool
+}
+
+// NewUnitState returns a new UnitState struct.
+func NewUnitState() *UnitState {
+	return &UnitState{}
+}
+
+// NoChanges returns true if no values of the struct have been set.
+func (u *UnitState) NoChanges() bool {
+	return !(u.relationStateSet || u.storageStateSet || u.stateSet || u.uniterStateSet)
+}
+
+// SetState sets the state value.
+func (u *UnitState) SetState(state map[string]string) {
+	u.stateSet = true
+	u.state = state
+}
+
+// State returns the unit's state and bool indicating
+// whether the data was set.
+func (u *UnitState) State() (map[string]string, bool) {
+	return u.state, u.stateSet
+}
+
+// SetUniterState sets the uniter state value.
+func (u *UnitState) SetUniterState(state string) {
+	u.uniterStateSet = true
+	u.uniterState = state
+}
+
+// UniterState returns the uniter state and bool indicating
+// whether the data was set.
+func (u *UnitState) UniterState() (string, bool) {
+	return u.uniterState, u.uniterStateSet
+}
+
+// SetRelationState sets the relation state value.
+func (u *UnitState) SetRelationState(state map[int]string) {
```

Comment:
Do we care if the caller passes a nil map here? Guess not; we will still treat that as a modification, right?

Reply:
Correct, if you add a nil map, it will delete all the data.  If you don't call it, nothing will be done to relation data.